File: src/currencies.py
from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class CurrencyDefinition:
    key: str
    label: str
    template_name: str
    trade_id: str
    official_name: str


CURRENCIES: tuple[CurrencyDefinition, ...] = (
    CurrencyDefinition("transmutation", "蜕变石", "currency_transmutation", "transmute", "Orb of Transmutation"),
    CurrencyDefinition("augmentation", "增幅石", "currency_augmentation", "aug", "Orb of Augmentation"),
    CurrencyDefinition("alteration", "改造石", "currency_alteration", "alt", "Orb of Alteration"),
    CurrencyDefinition("chance", "机会石", "currency_chance", "chance", "Orb of Chance"),
    CurrencyDefinition("alchemy", "点金石", "currency_alchemy", "alch", "Orb of Alchemy"),
    CurrencyDefinition("chaos", "混沌石", "currency_chaos", "chaos", "Chaos Orb"),
    CurrencyDefinition("regal", "富豪石", "currency_regal", "regal", "Regal Orb"),
    CurrencyDefinition("scouring", "重铸石", "currency_scouring", "scour", "Orb of Scouring"),
    CurrencyDefinition("exalted", "崇高石", "currency_exalted", "exalted", "Exalted Orb"),
    CurrencyDefinition("divine", "神圣石", "currency_divine", "divine", "Divine Orb"),
    CurrencyDefinition("annulment", "剥离石", "currency_annulment", "annul", "Orb of Annulment"),
    CurrencyDefinition("vaal", "瓦尔宝珠", "currency_vaal", "vaal", "Vaal Orb"),
    CurrencyDefinition("fusing", "链接石", "currency_fusing", "fusing", "Orb of Fusing"),
    CurrencyDefinition("jewellers", "珠宝匠之石", "currency_jewellers", "jewellers", "Jeweller's Orb"),
    CurrencyDefinition("chromatic", "幻色石", "currency_chromatic", "chrome", "Chromatic Orb"),
    CurrencyDefinition("gemcutter", "宝石匠棱镜", "currency_gemcutter", "gcp", "Gemcutter's Prism"),
    CurrencyDefinition("chisel", "制图石", "currency_chisel", "chisel", "Cartographer's Chisel"),
    CurrencyDefinition("bauble", "玻璃匠棱镜", "currency_bauble", "bauble", "Glassblower's Bauble"),
    CurrencyDefinition("whetstone", "磨刀石", "currency_whetstone", "whetstone", "Blacksmith's Whetstone"),
    CurrencyDefinition("scrap", "护甲片", "currency_scrap", "scrap", "Armourer's Scrap"),
    CurrencyDefinition("blessed", "祝福石", "currency_blessed", "blessed", "Blessed Orb"),
    CurrencyDefinition("regret", "后悔石", "currency_regret", "regret", "Orb of Regret"),
    CurrencyDefinition("wisdom", "知识卷轴", "currency_wisdom", "wisdom", "Scroll of Wisdom"),
    CurrencyDefinition("portal", "传送卷轴", "currency_portal", "portal", "Portal Scroll"),
    CurrencyDefinition("mirror", "镜像", "currency_mirror", "mirror", "Mirror of Kalandra"),
)

CURRENCY_BY_LABEL = {currency.label: currency for currency in CURRENCIES}
CURRENCY_BY_TEMPLATE = {currency.template_name: currency for currency in CURRENCIES}

# 国服字段偶尔会在汉字之间插入空格，例如“堆 叠 数 量”。
# 同时兼容英文客户端，便于单元测试和以后切换语言。
CURRENCY_STACK_COUNT_RE = re.compile(
    r"(?:堆\s*叠\s*数\s*量|Stack\s+Size)\s*[:：]\s*([\d,，]+)"
    r"(?:\s*/\s*[\d,，]+)?",
    re.IGNORECASE,
)


def currency_label(template_name: str) -> str:
    definition = CURRENCY_BY_TEMPLATE.get(template_name)
    return definition.label if definition is not None else template_name


def currency_template(label: str) -> str:
    definition = CURRENCY_BY_LABEL.get(label)
    return definition.template_name if definition is not None else ""


def currency_stack_count(copied_text: str) -> int | None:
    """从通货 Ctrl+C 文本中读取当前堆叠数量。"""

    match = CURRENCY_STACK_COUNT_RE.search(copied_text or "")
    if match is None:
        return None
    # 国服通货页的大堆叠会复制为“1,808 / 20”；斜杠前才是
    # 实际剩余量，逗号只是千位分隔符。
    raw_count = match.group(1).replace(",", "").replace("，", "")
    return int(raw_count)

File: electron/engine/currencies.ts
export interface CurrencyDefinition {
  key: string;
  label: string;
  templateName: string;
  tradeId: string;
  officialName: string;
}

export const CURRENCIES: CurrencyDefinition[] = [
  {
    key: "transmutation",
    label: "蜕变石",
    templateName: "currency_transmutation",
    tradeId: "transmute",
    officialName: "Orb of Transmutation",
  },
  {
    key: "augmentation",
    label: "增幅石",
    templateName: "currency_augmentation",
    tradeId: "aug",
    officialName: "Orb of Augmentation",
  },
  {
    key: "alteration",
    label: "改造石",
    templateName: "currency_alteration",
    tradeId: "alt",
    officialName: "Orb of Alteration",
  },
  {
    key: "chance",
    label: "机会石",
    templateName: "currency_chance",
    tradeId: "chance",
    officialName: "Orb of Chance",
  },
  {
    key: "alchemy",
    label: "点金石",
    templateName: "currency_alchemy",
    tradeId: "alch",
    officialName: "Orb of Alchemy",
  },
  {
    key: "chaos",
    label: "混沌石",
    templateName: "currency_chaos",
    tradeId: "chaos",
    officialName: "Chaos Orb",
  },
  {
    key: "regal",
    label: "富豪石",
    templateName: "currency_regal",
    tradeId: "regal",
    officialName: "Regal Orb",
  },
  {
    key: "scouring",
    label: "重铸石",
    templateName: "currency_scouring",
    tradeId: "scour",
    officialName: "Orb of Scouring",
  },
  {
    key: "exalted",
    label: "崇高石",
    templateName: "currency_exalted",
    tradeId: "exalted",
    officialName: "Exalted Orb",
  },
  {
    key: "divine",
    label: "神圣石",
    templateName: "currency_divine",
    tradeId: "divine",
    officialName: "Divine Orb",
  },
  {
    key: "annulment",
    label: "剥离石",
    templateName: "currency_annulment",
    tradeId: "annul",
    officialName: "Orb of Annulment",
  },
  {
    key: "vaal",
    label: "瓦尔宝珠",
    templateName: "currency_vaal",
    tradeId: "vaal",
    officialName: "Vaal Orb",
  },
  {
    key: "fusing",
    label: "链接石",
    templateName: "currency_fusing",
    tradeId: "fusing",
    officialName: "Orb of Fusing",
  },
  {
    key: "jewellers",
    label: "工匠石",
    templateName: "currency_jewellers",
    tradeId: "jewellers",
    officialName: "Jeweller's Orb",
  },
  {
    key: "chromatic",
    label: "幻色石",
    templateName: "currency_chromatic",
    tradeId: "chrome",
    officialName: "Chromatic Orb",
  },
  {
    key: "gemcutter",
    label: "宝石匠棱镜",
    templateName: "currency_gemcutter",
    tradeId: "gcp",
    officialName: "Gemcutter's Prism",
  },
  {
    key: "chisel",
    label: "制图石",
    templateName: "currency_chisel",
    tradeId: "chisel",
    officialName: "Cartographer's Chisel",
  },
  {
    key: "bauble",
    label: "玻璃匠棱镜",
    templateName: "currency_bauble",
    tradeId: "bauble",
    officialName: "Glassblower's Bauble",
  },
  {
    key: "whetstone",
    label: "磨刀石",
    templateName: "currency_whetstone",
    tradeId: "whetstone",
    officialName: "Blacksmith's Whetstone",
  },
  {
    key: "scrap",
    label: "护甲片",
    templateName: "currency_scrap",
    tradeId: "scrap",
    officialName: "Armourer's Scrap",
  },
  {
    key: "blessed",
    label: "祝福石",
    templateName: "currency_blessed",
    tradeId: "blessed",
    officialName: "Blessed Orb",
  },
  {
    key: "regret",
    label: "后悔石",
    templateName: "currency_regret",
    tradeId: "regret",
    officialName: "Orb of Regret",
  },
  {
    key: "wisdom",
    label: "知识卷轴",
    templateName: "currency_wisdom",
    tradeId: "wisdom",
    officialName: "Scroll of Wisdom",
  },
  {
    key: "portal",
    label: "传送卷轴",
    templateName: "currency_portal",
    tradeId: "portal",
    officialName: "Portal Scroll",
  },
  {
    key: "mirror",
    label: "镜像",
    templateName: "currency_mirror",
    tradeId: "mirror",
    officialName: "Mirror of Kalandra",
  },
];

export const CURRENCY_BY_LABEL = Object.fromEntries(CURRENCIES.map((c) => [c.label, c]));
export const CURRENCY_BY_TEMPLATE = Object.fromEntries(CURRENCIES.map((c) => [c.templateName, c]));

const CURRENCY_STACK_COUNT_RE =
  /(?:堆\s*叠\s*数\s*量|Stack\s+Size)\s*[:：]\s*([\d,，]+)(?:\s*\/\s*[\d,，]+)?/i;

export function currencyLabel(templateName: string): string {
  return CURRENCY_BY_TEMPLATE[templateName]?.label ?? templateName;
}

export function currencyStackCount(copiedText: string): number | null {
  const match = CURRENCY_STACK_COUNT_RE.exec(copiedText || "");
  if (!match) return null;
  return parseInt(match[1].replace(/[,，]/g, ""), 10);
}
